How to use the headless waitevent action?

### Initial thoughts
Discussing internally I got to know the following - 
This is the method, in go-rod lib causing this err:
```
// ParseMethodName to domain and name
func ParseMethodName(method string) (domain, name string) {
	arr := strings.Split(method, ".")
	return arr[0], arr[1]
}
```

### Discussed in https://github.com/projectdiscovery/nuclei/discussions/4411

<div type='discussions-op-text'>

<sup>Originally posted by **tovask** November 23, 2023</sup>
The [waitevent](https://docs.projectdiscovery.io/templates/protocols/headless#waitevent) action does not work for me, can somebody give me an example how to use it?

I tried with a simple template:
```yaml
id: waitevent
info:
    name: WaitEvent test
    severity: info
    author: Levente Kovats

headless:
  - steps:
      - action: navigate  
        args:
          url: "{{BaseURL}}"
      
      - action: waitevent
        args:
            event: 'Page.loadEventFired'
```
but Nuclei exited with an error:
```
$ ./nuclei -disable-update-check -headless -target https://imdb.com/ -templates waitevent.yaml -v -matcher-status

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.0.4

		projectdiscovery.io

[VER] Started metrics server at localhost:9092
[WRN] Template redirect_bug is not signed or tampered
[INF] Current nuclei version: v3.0.4 (outdated)
[INF] Current nuclei-templates version:  (outdated)
[INF] New templates added in latest release: 0
[INF] Templates loaded for current scan: 1
[INF] Targets loaded for current scan: 1
panic: runtime error: index out of range [1] with length 1

goroutine 24 [running]:
github.com/go-rod/rod/lib/proto.ParseMethodName(...)
	github.com/go-rod/rod@v0.114.0/lib/proto/a_interface.go:48
github.com/go-rod/rod.(*Browser).eachEvent(0xc0009c4550?, {0xc00050aa60, 0x20}, {0xc001c35278, 0x1, 0x0?})
	github.com/go-rod/rod@v0.114.0/browser.go:366 +0x4fe
github.com/go-rod/rod.(*Browser).waitEvent(0xc000ff8168?, {0xc00050aa60, 0x20}, {0x3211f40?, 0xc000ea7420?})
	github.com/go-rod/rod@v0.114.0/browser.go:348 +0x3a5
github.com/go-rod/rod.(*Page).WaitEvent(0xc00056e160, {0x3211f40, 0xc000ea7420})
	github.com/go-rod/rod@v0.114.0/page.go:564 +0x1b3
github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine.(*Page).WaitEvent(0xc00248a240, 0xc000fff920?, 0xc00248ab40?)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine/page_actions.go:574 +0x14b
github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine.(*Page).ExecuteActions(0xc00248a240, 0x16?, {0xc001b89220, 0x2, 0x0?}, 0x0?)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine/page_actions.go:72 +0x51b
github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine.(*Instance).Run(0xc001c0ef40, 0xc000d046c0, {0xc001b89220, 0x2, 0x2}, 0xc000da9a40, 0xc000d04798)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/engine/page.go:139 +0x846
github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless.(*Request).executeRequestWithPayloads(0xc0000b0300, 0xc000d046c0, 0x280167c?, 0x7?, 0xc000d046f0)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/request.go:142 +0x236
github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless.(*Request).ExecuteWithResults(0xc0000b0300, 0xc000ff81c8?, 0xc000da94a0, 0x0?, 0xc000da9680)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/headless/request.go:90 +0x637
github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults(0xc000fffe30, 0xc000d046a8, 0xc001c0ee80)
	github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic/exec.go:54 +0x2d3
github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute(0xc000d17d80, 0xc000d046a8)
	github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/exec.go:147 +0x203
github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1(0x1cf0687?, 0x0?, 0xc000da8a80)
	github.com/projectdiscovery/nuclei/v3/pkg/core/executors.go:128 +0x1d3
created by github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2 in goroutine 23
	github.com/projectdiscovery/nuclei/v3/pkg/core/executors.go:105 +0x4f1
```

Any idea?

Refs:
 - https://docs.projectdiscovery.io/templates/protocols/headless#waitevent
 - https://github.com/projectdiscovery/nuclei/discussions/1793
 - https://github.com/projectdiscovery/nuclei/blob/8d7bbdd97883b00e297b0cb5201e0516328eed38/pkg/protocols/headless/engine/page_actions.go#L553-L576
 - https://github.com/go-rod/rod/blob/167ecc0d37408d3a9ee8bb21bfc4bc8b4b336a86/page.go#L561-L565
 - https://github.com/go-rod/rod/blob/167ecc0d37408d3a9ee8bb21bfc4bc8b4b336a86/lib/proto/definitions.go#L803
 - https://github.com/go-rod/rod/blob/167ecc0d37408d3a9ee8bb21bfc4bc8b4b336a86/lib/proto/a_interface.go#L48</div>

--- a/cmd/integration-test/headless.go
+++ b/cmd/integration-test/headless.go
@@ -12,6 +12,7 @@ import (
 
 var headlessTestcases = []TestCaseInfo{
 	{Path: "protocols/headless/headless-basic.yaml", TestCase: &headlessBasic{}},
+	{Path: "protocols/headless/headless-waitevent.yaml", TestCase: &headlessBasic{}},
 	{Path: "protocols/headless/headless-self-contained.yaml", TestCase: &headlessSelfContained{}},
 	{Path: "protocols/headless/headless-header-action.yaml", TestCase: &headlessHeaderActions{}},
 	{Path: "protocols/headless/headless-extract-values.yaml", TestCase: &headlessExtractValues{}},

--- a/integration_tests/protocols/headless/headless-waitevent.yaml
+++ b/integration_tests/protocols/headless/headless-waitevent.yaml
@@ -1,0 +1,23 @@
+id: headless-waitevent
+
+info:
+    name: WaitEvent
+    severity: info
+    author: pdteam
+
+headless:
+  - steps:
+      # note waitevent must be used before navigating to any page
+      # unlike waitload
+      - action: waitevent
+        args:
+            event: 'Page.loadEventFired'
+  
+      - action: navigate  
+        args:
+          url: "{{BaseURL}}/"
+
+    matchers:
+      - type: word
+        words:
+          - "<html>"

--- a/integration_tests/protocols/headless/headless-waitevent.yaml
+++ b/integration_tests/protocols/headless/headless-waitevent.yaml
@@ -12,6 +12,7 @@ headless:
       - action: waitevent
         args:
             event: 'Page.loadEventFired'
+            max-duration: 15s
   
       - action: navigate  
         args:

--- a/pkg/protocols/headless/engine/page_actions.go
+++ b/pkg/protocols/headless/engine/page_actions.go
@@ -22,6 +22,7 @@ import (
 	"github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/utils/vardump"
 	protocolutils "github.com/projectdiscovery/nuclei/v3/pkg/protocols/utils"
 	httputil "github.com/projectdiscovery/nuclei/v3/pkg/protocols/utils/http"
+	contextutil "github.com/projectdiscovery/utils/context"
 	errorutil "github.com/projectdiscovery/utils/errors"
 	fileutil "github.com/projectdiscovery/utils/file"
 	folderutil "github.com/projectdiscovery/utils/folder"
@@ -46,7 +47,7 @@ func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, var
 	outData = make(map[string]string)
 	// waitFuncs are function that needs to be executed after navigation
 	// typically used for waitEvent
-	waitFuncs := make([]func(), 0)
+	waitFuncs := make([]func() error, 0)
 
 	// avoid any future panics caused due to go-rod library
 	defer func() {
@@ -63,7 +64,9 @@ func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, var
 				// if navigation successful trigger all waitFuncs (if any)
 				for _, waitFunc := range waitFuncs {
 					if waitFunc != nil {
-						waitFunc()
+						if err := waitFunc(); err != nil {
+							return nil, errorutil.NewWithErr(err).Msgf("error occurred while executing waitFunc")
+						}
 					}
 				}
 			}
@@ -88,7 +91,7 @@ func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, var
 		case ActionExtract:
 			err = p.ExtractElement(act, outData)
 		case ActionWaitEvent:
-			var waitFunc func()
+			var waitFunc func() error
 			waitFunc, err = p.WaitEvent(act, outData)
 			if waitFunc != nil {
 				waitFuncs = append(waitFuncs, waitFunc)
@@ -565,7 +568,7 @@ func (p *Page) ExtractElement(act *Action, out map[string]string) error {
 }
 
 // WaitEvent waits for an event to happen on the page.
-func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused parameter*/) (func(), error) {
+func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused parameter*/) (func() error, error) {
 	event := p.getActionArgWithDefaultValues(act, "event")
 	if event == "" {
 		return nil, errors.New("event not recognized")
@@ -581,9 +584,25 @@ func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused
 		return nil, errorutil.New("event %v is not a page event", event)
 	}
 	waitEvent = tmp
+	maxDuration := 10 * time.Second // 30 sec is max wait duration for any event
+
+	// allow user to specify max-duration for wait-event
+	if out["max-duration"] != "" {
+		var err error
+		maxDuration, err = time.ParseDuration(out["max-duration"])
+		if err != nil {
+			return nil, errorutil.NewWithErr(err).Msgf("could not parse max-duration")
+		}
+	}
 
 	// Just wait the event to happen
-	waitFunc := p.page.WaitEvent(waitEvent)
+	waitFunc := func() (err error) {
+		// execute actual wait event
+		ctx, cancel := context.WithTimeout(context.Background(), maxDuration)
+		defer cancel()
+		err = contextutil.ExecFunc(ctx, p.page.WaitEvent(waitEvent))
+		return
+	}
 	return waitFunc, nil
 }
 

--- a/pkg/protocols/headless/engine/page_actions.go
+++ b/pkg/protocols/headless/engine/page_actions.go
@@ -584,12 +584,12 @@ func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused
 		return nil, errorutil.New("event %v is not a page event", event)
 	}
 	waitEvent = tmp
-	maxDuration := 10 * time.Second // 30 sec is max wait duration for any event
+	maxDuration := 10 * time.Second // 10 sec is max wait duration for any event
 
 	// allow user to specify max-duration for wait-event
-	if out["max-duration"] != "" {
+	if value := p.getActionArgWithDefaultValues(act, "max-duration"); value != "" {
 		var err error
-		maxDuration, err = time.ParseDuration(out["max-duration"])
+		maxDuration, err = time.ParseDuration(value)
 		if err != nil {
 			return nil, errorutil.NewWithErr(err).Msgf("could not parse max-duration")
 		}

--- a/pkg/protocols/headless/engine/page_actions.go
+++ b/pkg/protocols/headless/engine/page_actions.go
@@ -43,6 +43,7 @@ const (
 
 // ExecuteActions executes a list of actions on a page.
 func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, variables map[string]interface{}) (outData map[string]string, err error) {
+	outData = make(map[string]string)
 	// waitFuncs are function that needs to be executed after navigation
 	// typically used for waitEvent
 	waitFuncs := make([]func(), 0)
@@ -652,10 +653,6 @@ func selectorBy(selector string) rod.SelectorType {
 	default:
 		return rod.SelectorTypeText
 	}
-}
-
-func (p *Page) getActionArg(action *Action, arg string) string {
-	return p.getActionArgWithValues(action, arg, nil)
 }
 
 func (p *Page) getActionArgWithDefaultValues(action *Action, arg string) string {

--- a/pkg/protocols/headless/engine/page_actions.go
+++ b/pkg/protocols/headless/engine/page_actions.go
@@ -42,14 +42,18 @@ const (
 )
 
 // ExecuteActions executes a list of actions on a page.
-func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, variables map[string]interface{}) (map[string]string, error) {
-	outData := make(map[string]string)
-
+func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, variables map[string]interface{}) (outData map[string]string, err error) {
 	// waitFuncs are function that needs to be executed after navigation
 	// typically used for waitEvent
 	waitFuncs := make([]func(), 0)
 
-	var err error
+	// avoid any future panics caused due to go-rod library
+	defer func() {
+		if r := recover(); r != nil {
+			err = errorutil.New("panic on headless action: %v", r)
+		}
+	}()
+
 	for _, act := range actions {
 		switch act.ActionType.ActionType {
 		case ActionNavigate:

--- a/pkg/protocols/headless/engine/page_actions.go
+++ b/pkg/protocols/headless/engine/page_actions.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"os"
 	"path/filepath"
+	"reflect"
 	"strconv"
 	"strings"
 	"sync"
@@ -43,11 +44,24 @@ const (
 // ExecuteActions executes a list of actions on a page.
 func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, variables map[string]interface{}) (map[string]string, error) {
 	outData := make(map[string]string)
+
+	// waitFuncs are function that needs to be executed after navigation
+	// typically used for waitEvent
+	waitFuncs := make([]func(), 0)
+
 	var err error
 	for _, act := range actions {
 		switch act.ActionType.ActionType {
 		case ActionNavigate:
 			err = p.NavigateURL(act, outData, variables)
+			if err == nil {
+				// if navigation successful trigger all waitFuncs (if any)
+				for _, waitFunc := range waitFuncs {
+					if waitFunc != nil {
+						waitFunc()
+					}
+				}
+			}
 		case ActionScript:
 			err = p.RunScript(act, outData)
 		case ActionClick:
@@ -69,7 +83,11 @@ func (p *Page) ExecuteActions(input *contextargs.Context, actions []*Action, var
 		case ActionExtract:
 			err = p.ExtractElement(act, outData)
 		case ActionWaitEvent:
-			err = p.WaitEvent(act, outData)
+			var waitFunc func()
+			waitFunc, err = p.WaitEvent(act, outData)
+			if waitFunc != nil {
+				waitFuncs = append(waitFuncs, waitFunc)
+			}
 		case ActionFilesInput:
 			if p.options.Options.AllowLocalFileAccess {
 				err = p.FilesInput(act, outData)
@@ -541,38 +559,27 @@ func (p *Page) ExtractElement(act *Action, out map[string]string) error {
 	return nil
 }
 
-type protoEvent struct {
-	event string
-}
-
-// ProtoEvent returns the cdp.Event.Method
-func (p *protoEvent) ProtoEvent() string {
-	return p.event
-}
-
 // WaitEvent waits for an event to happen on the page.
-func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused parameter*/) error {
+func (p *Page) WaitEvent(act *Action, out map[string]string /*TODO review unused parameter*/) (func(), error) {
 	event := p.getActionArgWithDefaultValues(act, "event")
 	if event == "" {
-		return errors.New("event not recognized")
+		return nil, errors.New("event not recognized")
 	}
-	protoEvent := &protoEvent{event: event}
 
-	// Uses another instance in order to be able to chain the timeout only to the wait operation
-	pageCopy := p.page
-	timeout := p.getActionArg(act, "timeout")
-	if timeout != "" {
-		ts, err := strconv.Atoi(timeout)
-		if err != nil {
-			return errors.Wrap(err, "could not get timeout")
-		}
-		if ts > 0 {
-			pageCopy = p.page.Timeout(time.Duration(ts) * time.Second)
-		}
+	var waitEvent proto.Event
+	gotType := proto.GetType(event)
+	if gotType == nil {
+		return nil, errorutil.New("event %v does not exist", event)
 	}
+	tmp, ok := reflect.New(gotType).Interface().(proto.Event)
+	if !ok {
+		return nil, errorutil.New("event %v is not a page event", event)
+	}
+	waitEvent = tmp
+
 	// Just wait the event to happen
-	pageCopy.WaitEvent(protoEvent)()
-	return nil
+	waitFunc := p.page.WaitEvent(waitEvent)
+	return waitFunc, nil
 }
 
 // pageElementBy returns a page element from a variety of inputs.

